src/hooks/useContractNames.ts

```diff
@@ -0,0 +1,15 @@
+import { useSnapshot } from 'valtio'
+import ContractMetadataStore from 'stores/ContractMetadataStore'
+import defaultProvider from 'helpers/providers/defaultProvider'
+
+export default function (addresses: string[]) {
+  const result = {} as { [address: string]: string | undefined }
+  for (const address of addresses) {
+    const { contractNames } = useSnapshot(ContractMetadataStore)
+    if (!contractNames[address]) {
+      ContractMetadataStore.fetchContractName(address, defaultProvider)
+    }
+    result[address] = contractNames[address]
```

Comment:
What will happen if the contract won't have a name? Won't it lead to an infinity loop

Reply:
Yes, you probably need to save `address` if there's no `contractNames[address]` after 1-2 attempts (you should try to fetch it 1-2 times)

Reply:
I tested it without a name, and it didn't caused an infite loop, it just showed up an actual address instead of a name. And I don't think that it's a good idea to try it with multiple attempts, I don't quite get it. If there's a name, we'll get it. If there's no, why should we try multiple times?